### PR TITLE
Proxito V2

### PR DESCRIPTION
**Pull request by stsewd**
This is an initial usage of the new proxito implementation to serve files. It's under a feature flag, so it can be enabled per-project.

- We still need to adapt our 404 handler to make use of this new implementation (it could be simplified with https://github.com/readthedocs/readthedocs.org/issues/10061)
- We are still using our re paths to match the URLs, but we don't make use of the captured parameters, we use the unresolver for that. This also means that things like https://github.com/readthedocs/readthedocs.org/issues/2292 won't be solved till we do the whole migration.
-  There is a lot of repetition from the original get method, but some of it could be simplified with https://github.com/readthedocs/readthedocs.org/issues/10065.

Tests will pass once we have merged one private PR that we have pending.

> **Note**
> We don't support custom urlconfs yet, so we shouldn't enable it for those projects yet.

This is on top of https://github.com/readthedocs/readthedocs.org/pull/10037.

### CODE REVIEW

File: readthedocs/proxito/views/serve.py
Comment (stsewd) [inline readthedocs/proxito/views/serve.py:346]: I think I can just normalize the path before returning it here, so we would just need to check for `/`. If we want to original value, we have it in the `path` variable.

File: readthedocs/proxito/tests/test_full.py
Comment (humitos) [inline readthedocs/proxito/tests/test_full.py:493]: Nice trick! 💯

Comment (humitos) [reply readthedocs/proxito/tests/test_full.py:493]: We should probably add a `# TODO` comment here to come back in the future and delete these classes.

File: readthedocs/proxito/views/serve.py
Comment (humitos) [inline readthedocs/proxito/views/serve.py:311]: What if "the middleware" is the one that calls `unresolver.unresolve_path()` and inject it in the `request` instead? That way, it would be available for everybody who needs it.

Probably not a clean implementation or good idea, but I just thought about that and wrote it down just in case 🙃

File: readthedocs/proxito/views/serve.py
Comment (humitos) [inline readthedocs/proxito/views/serve.py:313]: I'm not sure to understand _why_ the flow for the redirects is managed under `UnresolvedPathError` but I don't like too much this flow.

The name of the exception is confusing. We are raising an error, but it's a common/normal/success path to follow for all canonical and system redirects. So, I'd not use the `Error` word to handle this flow.

If you are handling these cases under a try/except clause on purpose, I'd suggest at least, to change the name of the exception to `UnresolvedPathNotMatched`, meaning that it didn't match any of the regular path patterns.

File: readthedocs/proxito/views/serve.py
Comment (humitos) [inline readthedocs/proxito/views/serve.py:405]: This code was already under the try/except clause. Why are we handling canonical redirects here as well? What are the different cases that each of them are handling?

File: readthedocs/proxito/views/serve.py
Comment (humitos) [inline readthedocs/proxito/views/serve.py:442]: We should log a warning here.

**Review by humitos — COMMENTED** [review]
This is looking good 👍🏼  I asked for some explanation about the flow because I'm not sure to fully follow it. I may have more feedback once I get the replies I'm looking for.

File: readthedocs/proxito/views/serve.py
Comment (stsewd) [reply readthedocs/proxito/views/serve.py:311]: That would require moving the exception control logic to the middleware, and we want to do that only in the docs view.

File: readthedocs/proxito/views/serve.py
Comment (stsewd) [reply readthedocs/proxito/views/serve.py:313]: That's an invalid path for versioned documentation, we don't serve files at `docs.rtd.io/` if the project isn't a single version, so being an error makes sense to me.

In proxito, when that happens, we want to redirect to the main version, we don't want to 404.

File: readthedocs/proxito/views/serve.py
Comment (stsewd) [reply readthedocs/proxito/views/serve.py:405]: Our current implementation always does canonical redirects before checking for any other redirects or serving files, so we need it at both places, this won't be required with https://github.com/readthedocs/readthedocs.org/issues/10065.

File: readthedocs/proxito/views/serve.py
Comment (stsewd) [reply readthedocs/proxito/views/serve.py:442]: We already have a warning in the function itself https://github.com/readthedocs/readthedocs.org/blob/d499203f6edf0106f146941ac6c120d831e916e8/readthedocs/proxito/views/mixins.py#L456-L459

File: readthedocs/proxito/views/serve.py
Comment (humitos) [reply readthedocs/proxito/views/serve.py:324]: > That's an invalid path for versioned documentation, we don't serve files at docs.rtd.io/ if the project isn't a single version, so being an error makes sense to me.

In that case, the name of the extension could make some reference to this. The current name is confusing to me and does not communicate exactly what's the error.

If I understand correctly the case you are describing, this name would make sense `AttemptToServeFilesFromRootOnVersionedDocs`. Of course, it's too long. However, if I'm right with my interpretation of the exception, we should find a better name that communicates this clearly.

**Review by humitos — APPROVED** [review]
I'm a bit confused with so many refactor PRs for El Proxito. In particular because when I comment something, the thing I'm saying may be happening in another PR at the same time or it _will be done_ in the sooner future.

I think that's fine and I'm happy we are moving forward with this new implementation. However, sometimes I'm having problems to understand _the big picture_ with all these changes happening at the same time. Because of this, I'm pinging @ericholscher for a review here as well, just to have a second pair of eyes just in case -- but I think it's fine to merge without his approval, tho.

I suggested to keep working on some exceptions' names to make them easy to follow/understand at first sight without requiring having all the code in mind.

File: readthedocs/proxito/views/serve.py
Comment (ericholscher) [inline readthedocs/proxito/views/serve.py:308]: I don't really understand this TODO?

File: readthedocs/proxito/views/serve.py
Comment (ericholscher) [reply readthedocs/proxito/views/serve.py:324]: I agree with @humitos that it's odd to have so much logic in an exception handler. I think we likely want another way to represent this, since it's actually the majority of the code here?

> In proxito, when that happens, we want to redirect to the main version, we don't want to 404.

I'm also not 100% sure this is correct. Automatic redirects have bitten us in the past, so I'd like to make sure we have a plan and documentation for what these redirects do. 

Is the idea that `docs.rtd.io/index.html` will redirect the same as `docs.rtd.io/page/index.html`?

File: readthedocs/proxito/views/serve.py
Comment (ericholscher) [inline readthedocs/proxito/views/serve.py:358]: I don't understand this logic. Comment could be better.

File: readthedocs/proxito/views/serve.py
Comment (ericholscher) [inline readthedocs/proxito/views/serve.py:363]: This is new behavior, right? I'm not 100% sold on this, especially given how it might be weird for changing to/from a single_version project.

File: readthedocs/proxito/views/serve.py
Comment (ericholscher) [reply readthedocs/proxito/views/serve.py:453]: Having these bubble up and be ignored in multiple places definitely feels odd. Should we actually be exposign the error to the user, since it's some kind of misconfiguration of redirects?

**Review by ericholscher — APPROVED** [review]
This is really exciting. I'm 👍 on shipping this since it's behind a feature flag, it's low risk and we can test it in production.

File: readthedocs/proxito/views/serve.py
Comment (humitos) [reply readthedocs/proxito/views/serve.py:453]: That's probably a better outcome actually. Is there a specific HTTP code response for this?

File: readthedocs/proxito/views/serve.py
Comment (stsewd) [reply readthedocs/proxito/views/serve.py:324]: > Is the idea that docs.rtd.io/index.html will redirect the same as docs.rtd.io/page/index.html?

Nope, this is `docs.rtd.io/` will redirect to `docs.rtd.io/en/latest` and `docs.rtd.io/projects/subproject/` will redirect to `docs.rtd.io/projects/subproject/en/latest/`

File: readthedocs/proxito/views/serve.py
Comment (stsewd) [reply readthedocs/proxito/views/serve.py:347]: This is how our current implementation works!

https://github.com/readthedocs/readthedocs.org/blob/c3e54ba77ed1857dcd4ba7efedc8d04f141b128a/readthedocs/proxito/views/serve.py#L171-L186

This doesn't affect single versions.

File: readthedocs/proxito/views/serve.py
Comment (stsewd) [reply readthedocs/proxito/views/serve.py]: Changed the name to `InvalidPathForVersionedProjectError`

File: readthedocs/proxito/views/serve.py
Comment (stsewd) [reply readthedocs/proxito/views/serve.py]: I have inverted this condition, I think that should make it clear

File: readthedocs/proxito/views/serve.py
Comment (ericholscher) [reply readthedocs/proxito/views/serve.py]: @stsewd thanks for the rewrite -- much clearer :)

File: readthedocs/proxito/views/serve.py
Comment (ericholscher) [reply readthedocs/proxito/views/serve.py]: Ah, gotcha. This is only at `/`, and not all files within, right?

Comment (stsewd) [reply readthedocs/proxito/views/serve.py]: yeah, only the root for non single version projects

File: readthedocs/proxito/views/serve.py
Comment (humitos) [reply readthedocs/proxito/views/serve.py]: This exception has a pretty explicit name 💯 . I may not even need context to have an idea what could be happening.

Comment (humitos) [reply readthedocs/proxito/views/serve.py]: > I agree with @humitos that it's odd to have so much logic in an exception handler. I think we likely want another way to represent this, since it's actually the majority of the code here?

Should we open an issue about this to consider other refactor alternatives maybe 🤔 ?

@stsewd how do you feel about our feedback? Do you also find hard to parse this code or you already have it in mind and feel it more natural?

Comment (stsewd) [reply readthedocs/proxito/views/serve.py]: Logic inside this exception is pretty small now

https://github.com/readthedocs/readthedocs.org/blob/776260a47165ca43b0dfb4833b78ae5556a22421/readthedocs/proxito/views/serve.py#L331-L354

But if you mean like in general, i.e. using multiple return types instead of exceptions. The first implementation I wrote was setting some values to `None` depending on the case, and the caller would need to check if a given value was `None` or not, I found that repetitive and not that explicit. Having a single return type makes it easier, since the unresolved path/URL is valid or not, there aren't "incomplete" results.

Comment (ericholscher) [reply readthedocs/proxito/views/serve.py]: I'm not super worried about this flow, and the more explicit exception makes it pretty clear to me. I don't know if it's even work another issue.

**Pull request merged**
